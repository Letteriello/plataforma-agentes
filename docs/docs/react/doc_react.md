Claro. Aqui está a documentação detalhada da arquitetura front-end, 100% em formato Markdown, utilizando React e a abordagem `shadcn/ui`.

# **Documentação da Arquitetura Front-end (React + shadcn/ui)**

## **1. Visão Geral e Filosofia**

O front-end desta plataforma será a interface principal para a poderosa engine do **ADK (Agent Development Kit)**. Nossa filosofia é criar uma UI limpa, acessível e altamente personalizável, utilizando a abordagem de componentização oferecida pelo **shadcn/ui**.

Não usaremos uma biblioteca de componentes tradicional. Em vez disso, teremos componentes que podemos possuir e adaptar, copiados diretamente para nossa base de código. Isso nos dá máxima flexibilidade para construir uma experiência única, permitindo que qualquer usuário possa:

- **Interagir** com agentes de forma fluida e conversacional.
- **Visualizar** o "raciocínio" do agente (ex: uso de ferramentas, status).
- **Configurar** o comportamento dos agentes através de formulários claros e intuitivos.
- **Gerenciar** sessões, agentes e seus recursos (artefatos).

---

## **2. Stack Tecnológica Principal**

A stack foi refinada para se alinhar com o ecossistema do shadcn/ui:

- **Framework:** **React 18+** (utilizando Hooks).
- **Linguagem:** **TypeScript** para segurança de tipos robusta, essencial para as complexas estruturas de dados do ADK.
- **Componentes:** **shadcn/ui**. Coleção de componentes reutilizáveis, lindamente projetados, que você pode instalar e possuir. Usaremos componentes como `Button`, `Input`, `Card`, `Dialog`, `Select`, `Textarea`, e `Alert`.
- **Estilização:** **Tailwind CSS**. Utilitário CSS que nos permite construir designs customizados rapidamente, sendo a base de estilização do shadcn/ui.
- **Gerenciamento de Estado:** **Zustand**. Leve, moderno e poderoso para gerenciar nosso estado global (sessão, agente ativo, etc.).
- **Roteamento:** **React Router**.
- **Comunicação com API:** **Axios** para requisições RESTful e a API nativa **`EventSource`** para a comunicação via streaming (SSE).

---

## **3. Estrutura de Pastas do Projeto**

A estrutura será otimizada para acomodar os componentes do shadcn/ui:

```
src/
|-- /app/                # Roteamento e páginas (se usando Next.js App Router)
|-- /assets/             # Imagens, fontes, etc.
|-- /components/         # Componentes específicos do nosso projeto
|   |-- /features/       # Agrupamento de componentes por funcionalidade (chat, config)
|   |-- /ui/             # LOCAL ONDE OS COMPONENTES DO SHADCN/UI SÃO INSTALADOS
|-- /hooks/              # Hooks personalizados (useAgentStream, useSession)
|-- /lib/                # Funções utilitárias, incluindo a `cn` do shadcn/ui
|-- /services/           # Lógica de comunicação com o backend
|-- /store/              # Configuração do Zustand
|-- /types/              # Definições de tipos TypeScript
|-- layout.tsx           # Layout principal
|-- page.tsx             # Página de entrada
```

---

## **4. Arquitetura de Componentes com shadcn/ui**

Mapeamos os conceitos do ADK para uma hierarquia de componentes React, utilizando o shadcn/ui.

### **`<ChatPage />`** (Página Principal)

Componente que orquestra a interface. Usará o componente `Resizable` do shadcn/ui para criar um layout com painéis redimensionáveis.

- **`<AgentSidebar />`**: Um `<ResizablePanel />` contendo uma lista de agentes. Cada agente pode ser um `<Button variant="ghost">` dentro de uma lista.
- **`<ChatWindow />`**: O painel principal (`<ResizablePanel />`).
    - **`<MessageList />`**: Um `div` com scroll que renderiza a lista de `Events` da sessão.
    - **`<Message />`**: Componente inteligente que usa `Card`, `Alert` e `Avatar` do shadcn/ui para renderizar diferentes tipos de eventos:
        - **Mensagem de Texto:** Um `div` estilizado com Tailwind CSS para a "bolha" de chat.
        - **Chamada de Ferramenta (`tool_code`):** Um componente `<Alert>` para destacar a ação. Ex: `<Alert><Terminal className="h-4 w-4" /> <AlertTitle>Executando Ferramenta</AlertTitle><AlertDescription>get_current_weather(...)</AlertDescription></Alert>`.
        - **Resultado de Ferramenta:** Pode ser exibido dentro de um `<Card>` com um `<CardContent>`.
        - **Artefato (`Artifact`):** Um `<Card>` com um `<Avatar>` para o ícone do tipo de arquivo e um `<Button>` para download/visualização, que pode abrir um `<Dialog>`.
    - **`<InputBar />`**: A barra de entrada na parte inferior.
        - Utiliza o `<Input />` e `<Button />` do shadcn/ui.
        - Um botão com um ícone de clipe usará um `<input type="file" hidden />` para upload de **artefatos**.

### **`<AgentConfigPage />`** (Página de Configuração)

Uma página de formulário construída inteiramente com componentes shadcn/ui para uma experiência de usuário limpa e consistente.

- **`<InstructionEditor />`**: Um `<Card>` contendo um `<Label htmlFor="instructions">Instruções do Agente</Label>` e um `<Textarea id="instructions" />` para o usuário definir o comportamento do agente.
- **`<ToolSelector />`**: Um `<Card>` com o título "Ferramentas Disponíveis". Dentro, uma lista de ferramentas, cada uma com um `<Checkbox />` e `<Label />` para habilitar/desabilitar.
- **`<ModelSelector />`**: Um `<Select />` para que o usuário escolha o `Model` (Gemini, Claude, etc.) para o `LlmAgent`.

---

## **5. Gerenciamento de Estado com Zustand**

Nosso store global (`src/store/sessionStore.ts`) gerenciará o estado da interação. Ele é agnóstico à biblioteca de UI.

TypeScript

```
// src/store/sessionStore.ts
import create from 'zustand';
import { Session, Agent } from '../types/adk.types';

interface SessionState {
  activeSession: Session | null;
  availableAgents: Agent[];
  activeAgent: Agent | null;
  isStreaming: boolean;
  setSession: (session: Session) => void;
  addEventToSession: (event: Event) => void;
  setActiveAgent: (agent: Agent) => void;
  setStreaming: (isStreaming: boolean) => void;
}

export const useSessionStore = create<SessionState>((set) => ({
  activeSession: null,
  availableAgents: [],
  activeAgent: null,
  isStreaming: false,
  setSession: (session) => set({ activeSession: session }),
  addEventToSession: (event) => set((state) => ({
    activeSession: state.activeSession
      ? { ...state.activeSession, events: [...state.activeSession.events, event] }
      : null,
  })),
  setActiveAgent: (agent) => set({ activeAgent: agent }),
  setStreaming: (isStreaming) => set({ isStreaming }),
}));
```

---

## **6. Comunicação com o Backend (API & Streaming)**

A camada de serviço é responsável pela comunicação de dados.

- **`apiService.ts`**: Continuará usando Axios para chamadas REST.
- **`useAgentStream.ts`**: Nosso hook personalizado usará `EventSource` para gerenciar a conexão de streaming SSE, permitindo a comunicação em tempo real entre o agente e o cliente.

---

## **7. Traduzindo Conceitos ADK para a UI (com shadcn/ui)**

|   |   |
|---|---|
|**Conceito do ADK**|**Implementação no Front-end (React + shadcn/ui)**|
|**Session** & **Events**|O estado `activeSession` no **Zustand**. O componente **`<MessageList />`** renderiza `Events` usando **`<Card>`** e **`<Alert>`** para diferentes tipos de mensagem.|
|**Streaming**|O hook **`useAgentStream`** com `EventSource`. O texto dentro do componente `<Message />` é atualizado token a token, criando um efeito de digitação em tempo real.|
|**Tools** & **Function Calling**|O componente **`<Alert>`** é usado para notificar visualmente que uma ferramenta está em uso, com o nome da ferramenta e seus parâmetros.|
|**Artifacts**|O upload é feito via **`<Input type="file">`** no `<InputBar />`. Artefatos são exibidos em um **`<Card>`**, e a visualização completa pode ser feita em um **`<Dialog>`**.|
|**LLM, Workflow Agents**|Na **`<AgentSidebar />`**, cada agente é um **`<Button variant="ghost">`**. Podemos usar ícones do `lucide-react` para diferenciar os tipos de agente.|
|**RunConfig** & **Callbacks**|A UI pode exibir um ícone de carregamento diferente se o streaming estiver desabilitado (`RunConfig`). Um `Callback` de segurança pode resultar em um `<Alert variant="destructive">` com a mensagem de bloqueio.|
|**Authentication for Tools**|A página de configuração terá campos **`<Input type="password">`** e **`<Label>`** para o usuário inserir credenciais de forma segura, que serão enviadas para o backend.|

---

## **8. Conclusão**

Adotar **shadcn/ui** e **Tailwind CSS** nos posiciona na vanguarda do desenvolvimento front-end. Ganhamos uma velocidade de desenvolvimento incrível e controle total sobre a aparência e o comportamento da nossa aplicação, sem sacrificar a qualidade ou a acessibilidade. Esta arquitetura é a base ideal para construir a plataforma de agentes autônomos que você idealizou, garantindo que ela seja não apenas poderosa, mas também bonita e extremamente agradável de usar.