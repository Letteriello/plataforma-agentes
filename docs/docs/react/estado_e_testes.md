# Guia de Gerenciamento de Estado, Testes e Arquitetura Desacoplada

Este documento resume como o projeto utiliza **Zustand** para gerenciar estado, **Vitest** com **React Testing Library** para testes e a filosofia de desacoplamento de UI por meio de **hooks** e **serviços**.

## 1. Gerenciamento de Estado com Zustand

### 1.1 Documentação Oficial
- Repositório: <https://github.com/pmndrs/zustand>
- Guia rápido: consulte o README do projeto para criar seu primeiro store.

### 1.2 Melhores Práticas
- **Divisão de Stores**: cada domínio possui seu store, como `agentStore` e `sessionStore`, evitando um estado global gigantesco.
- **Seletores**: extraia apenas o que o componente precisa para reduzir re-renderizações.
- **Ações**: defina funções dentro do `create` para atualizar o estado de forma centralizada.

### 1.3 Exemplo no Projeto
O arquivo `client/src/store/agentStore.ts` define um store com lista de agentes e agente ativo, incluindo ações como `addAgent` e `setActiveAgent` para manipular o estado.

### 1.4 Alternativas
- **Redux Toolkit** – robusto e com ótimas ferramentas de depuração, porém mais verboso.
- **Jotai** – abordagem atômica e simples, mas pode ser limitado em cenários complexos.

## 2. Testes com Vitest e React Testing Library

### 2.1 Vitest
- Configurado em `client/vite.config.ts` dentro da chave `test`.
- `client/src/setupTests.ts` importa `@testing-library/jest-dom` para matchers adicionais.

### 2.2 React Testing Library
- A filosofia é testar o comportamento visível para o usuário.
- Use seletores acessíveis (`getByRole`, `getByLabelText`) antes de recorrer a `getByTestId`.

### 2.3 Exemplo no Projeto
O teste `client/src/components/agents/AgentList.test.tsx` mostra como mockar o store do Zustand e verificar a renderização de agentes com `screen.getByText()`.

## 3. Arquitetura de Desacoplamento

### 3.1 Camada de Serviços
`client/src/api/agentService.ts` exemplifica a separação das chamadas de API. Nenhum componente de UI faz requisições diretamente.

### 3.2 Hooks Personalizados
Hooks como `client/src/hooks/useTheme.ts` encapsulam lógica reutilizável e servem de ponte entre serviços ou stores e os componentes.

A UI concentra-se apenas em renderizar e delega efeitos colaterais e comunicação de dados aos hooks e serviços.
