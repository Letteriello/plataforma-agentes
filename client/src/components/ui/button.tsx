import { type VariantProps } from 'class-variance-authority'
import { Loader2 } from 'lucide-react'
import * as React from 'react'

import { cn } from '@/lib/utils'

import { buttonVariants } from './button-variants'

export interface ButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement>,
    VariantProps<typeof buttonVariants> {
  isLoading?: boolean
  leftIcon?: React.ReactNode
  rightIcon?: React.ReactNode
}

const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  (
    {
      className,
      variant,
      size,
      
      isLoading = false,
      disabled = false,
      children,
      leftIcon,
      rightIcon,
      ...props
    },
    ref,
  ) => {
    // Removida dependÃªncia do Slot do Radix UI
    return (
      <button
        className={cn(buttonVariants({ variant, size, className }))}
        ref={ref}
        disabled={isLoading || disabled}
        {...props}
      >
        {isLoading ? (
          <>
            <Loader2 className="mr-2 h-4 w-4 animate-spin" />
            {children}
          </>
        ) : (
          <>
            {leftIcon && <span className="mr-2">{leftIcon}</span>}
            {children}
            {rightIcon && <span className="ml-2">{rightIcon}</span>}
          </>
        )}
      </button>
    )
  },
)
Button.displayName = 'Button'

export { Button, buttonVariants }
