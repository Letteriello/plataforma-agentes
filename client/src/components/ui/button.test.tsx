import { render } from '@testing-library/react';
import { Button } from './button';

describe('Button component', () => {
  it('matches snapshot', () => {
    const { container } = render(<Button>Click</Button>);
    expect(container).toMatchSnapshot();
  });
});
