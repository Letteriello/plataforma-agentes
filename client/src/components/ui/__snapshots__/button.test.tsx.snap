// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button component > matches snapshot 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none ring-offset-background bg-primary text-primary-foreground hover:bg-primary/90 h-10 py-2 px-4"
  >
    Click
  </button>
</div>
`;
